texlive_latest: init at 2023.20230606

apfelkuchen6 opened the pull request:
###### Description of changes

This is an alternative approach to #222142 suggested by @xworld21 in https://github.com/NixOS/nixpkgs/pull/222142#issuecomment-1577449380: This adds texlive-2023 in addition to the old version instead of replacing it.

I also noticed the following:
- the procedure described in UPGRADING.md is no longer applicable since `texlive.scheme-full.pkgs` is empty since #218331. As this will be resolved/changed by #223104 anyway, I haven't bothered with updating the instructions yet.
- `dvisvgm`'s pdf conversion feature is incompatible with `ghostscript >= 10`. We have a test for that, but the test isn't good enough: It just successfully creates an empty svg file. The `dvisvgm` version shipped with texlive-2023 also has a `mupdf`-backend as a workaround. I'm not sure what to do about the old version.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

drupol commented on pkgs/test/texlive/default.nix:
```diff
@@ -46,10 +46,23 @@
+    # the return code now indicates whether warnings/errors have been emmited.
+    # chktex is supposed to return 2 when it (successfully) finds warnings, but no errors,
+    # see http://git.savannah.nongnu.org/cgit/chktex.git/commit/?id=ec0fb9b58f02a62ff0bfec98b997208e9d7a5998
+    if texlive.bin.core.version >= "2023"
```
```suggestion
    if (lib.versionAtLeast "2023" texlive.bin.core.version)
```

drupol commented on pkgs/tools/typesetting/tex/luametatex/default.nix:
```diff
@@ -0,0 +1,23 @@
+{ stdenv, lib, fetchFromGitHub, cmake, ... }:
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

drupol commented on pkgs/tools/typesetting/tex/texlive/bin.nix:
```suggestion
  buildInputs = core.buildInputs ++ [ core cairo harfbuzz icu graphite2 libX11 ] ++ lib.optional (lib.versionAtLeast "2023" version) potrace;
```

drupol reviewed (commented):
Just added a few comments while passing by :)

xworld21 commented on pkgs/test/texlive/default.nix:
```diff
@@ -1,8 +1,8 @@
 {
 
-  tlpdbNix = runCommand "texlive-test-tlpdb-nix" {
+  tlpdbNix = if isLatest then null else runCommand "texlive-test-tlpdb-nix" {
```
Once the rest of the PR has stabilized: `isLatest` should be replaced by an appropriate test that checks `latest-diff.nix`.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/bin.nix:
How hard would it be to have `bin.nix` as before, and an extra `bin-latest.nix` with overrides only? Which I believe would be the more nixpkgs-y way of providing two versions.

Alternatively, `bin-latest.nix` could be a straight copy of `bin.nix`, which would become the new `bin.nix` once it becomes frozen.

apfelkuchen6 replied:
Exactly as before is rather inconvenient as the sources would have to be overridden for every derivation. Just passing in the source as an argument and doing the patches and version-specific hacks via overrides shouldn't be too hard at this time as 2023 mostly just adds hacks over 2022. This might become more inconvenient if parts of hacks have to be removed.

I don't really like either approach. Using overrides makes upgrading less convenient and having a complete copy of the file feels really wasteful.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -317,23 +317,29 @@ in (buildEnv {
     wrapBin
-  '' +
+  ''
+  + bin.cleanBrokenLinks +
```
(Incidental: `cleanBrokenLinks` will disappear with the bin container patch. I believe it is already a no-op since we started using `texlinks`.)

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -317,23 +317,29 @@ in (buildEnv {
       FORCE_SOURCE_DATE=1 TZ= faketime -f '@1980-01-01 00:00:00 x0.001' luatex --luaonly mtxrun.lua --generate
+
+      if [[ -e "$out/bin/luametatex" ]]; then
+        # we intentionally don't use the scripts.lst-mechanism here since we don't want to strip the suffix or wrap the scripts here.
```
(This should also become unnecessary after the bin container patch.)

xworld21 commented on pkgs/tools/typesetting/tex/texlive/default.nix:
This file should then be called `make-tex-live.nix` or similar, probably. And possibly in the future be exposed as `makeTeXLive`, once the interface is sufficiently stable.

xworld21 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5003,7 +5005,8 @@ with pkgs;
 
   # TeX Live; see https://nixos.org/nixpkgs/manual/#sec-language-texlive
-  texlive = recurseIntoAttrs (callPackage ../tools/typesetting/tex/texlive { });
+  texlive = recurseIntoAttrs (callPackage ../tools/typesetting/tex/texlive/frozen.nix { });
```
Picking up on the above, I believe tradition is to use `default.nix` for the stable version (thus `default.nix -> make-tex-live.nix` and `frozen.nix -> default.nix`).

apfelkuchen6 replied:
There already are a lot of files here and I feel like this makes it even less obvious where what is as this breaks up the symmetry between the two versions and replaces `default.nix` with a mostly empty file just containing version information. How do you feel about the following layout:

```
├── bin.nix
├── combine.nix
├── default.nix
├── generate-fixed-hashes.nix
├── latest
│   ├── fixed-hashes.nix
│   ├── tlpdb.nix
│   └── version-info.nix
├── make-texlive-packages.nix
├── stable
│   ├── fixed-hashes.nix
│   ├── tlpdb.nix
│   └── version-info.nix
├── tl2nix.sed
└── UPGRADING.md
```

Clearly version specific stuff is put into subdirectories whereas `default.nix` becomes a mostly generic entry point that builds the texlive(_latest) scope. By default it imports the version-info from the stable subdirectory, but this is overridden in `all-packages.nix` for the latest version. The code that builds the texlive packages (currently `tl` in `default.nix` is extracted into a separate file `make-texlive-packages`.

xworld21 replied:
@apfelkuchen6 now that the binary containers are in, what do you think of refactoring the packages in `texlivePackages`? Presumably with `tlpdb.nix` and `fixed-hashes.nix` in a separate `texlive-packages/` folder, while `texlive` keeps the combine machinery in place. The Nix expressions to build the packages should probably stay under `texlive`, e.g. `texlive.buildPackage`. Priority should be to make the entire thing easy to override, something which is currently very hard.

You have a few pieces here already – shall we start from those? Say with a new 0-rebuild PR?

xworld21 commented on pkgs/tools/typesetting/tex/texlive/latest.nix:
```diff
@@ -0,0 +1,48 @@
+callPackage ./default.nix {
+  inherit version tlpdb tlpdbxz fixedHashes urlPrefixes src;
+  texlive = texlive_latest;
+}
```
E.g. could the 2023-specific overrides happen here? Likewise 2022-only patches could be specified in `frozen.nix`.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/latest.nix:
```diff
@@ -0,0 +1,48 @@
+  urlPrefixes = with version;
+    [
+      # CTAN mirror, not frozen, might result in misses
+      "https://ftp.rrze.uni-erlangen.de/ctan/systems/texlive/tlnet"
```
(We can do this now, because copy-tarballs will try every url until one works.)

xworld21 commented on pkgs/tools/typesetting/tex/texlive/frozen.nix:
```diff
@@ -0,0 +1,36 @@
+    ];
+
+  tlpdbxz = fetchurl {
+    urls = map (up: "${up}/tlpkg/texlive.tlpdb.xz") urlPrefixes;
```
The URL structure of a TeX Live repository is always the same, so this computation can be shifted to `default.nix` (so this would just be the hash of `tlpdbxz`).

apfelkuchen6 replied:
it actually is slightly different in `latest.nix` as we explicitly select the last mirror there. But this also works for the frozen.version, so it could be factored out

xworld21 reviewed (changes requested):
That's very good. I thought it would be a lot harder.

I like that `frozen.nix` and `latest.nix` more or less contain the essential information about each specific version. However, the logic is still duplicated around – `default.nix` has special code for 2022 and 2023; the urls can be computed once in `default.nix`, like before; and similar (see comments).

I suggest we push further, and have a `make-tex-live.nix` that takes as input: the version info, the tlpdb.xz hash (just the hash, not the url!), tlpdb.nix, the binary sources, and fixed hashes. Then `frozen.nix` calls that function and plugs in whatever overrides are required (e.g. version-specific patches).

xworld21 committed:
tests.texlive.fixedHashes: init

The assertion that all TeX Live packages have a fixed hash is time
consuming and should only be checked when running tests.

apfelkuchen6 committed:
texlive: move combinePkgs from default.nix to combine.nix

apfelkuchen6 committed:
texlive: make scope

apfelkuchen6 committed:
texlive: factor out tlpdb overrides

apfelkuchen6 committed:
texlive.combine: explicitly list params

This prepares the use of callPackage (to avoid repeating the parameter names all
over the place) and makes static analysis possible

apfelkuchen6 committed:
treewide: access texlive packages via texlive.pkgs instead of the toplevel

apfelkuchen6 committed:
luametatex: init at 2.10.08

apfelkuchen6 committed:
texlive: indent body of attrset

this is purely a whitespace change that was separated from the previous commit
to make the diff easier to read

apfelkuchen6 committed:
texlive: extract patches specific to texlive-2022 into stable.nix

apfelkuchen6 committed:
texlive.combine: prepare for texlive-2023

apfelkuchen6 committed:
texlive: extract generic interface

This generic interface allows the addition of texlive-2023 without repeating too
much code or case distinctions

TODO: rename default.nix -> generic.nix, stable.nix -> default.nix
This will be done later as this makes rebasing a lot harder

apfelkuchen6 committed:
texlive: fix cross

apfelkuchen6 committed:
texlive: adjust generate-fixed-hashes for use with texlive_latest

apfelkuchen6 committed:
texlive_latest: init at 2023.20230807
